rfcs/v3/authorization-rules.md

```diff
@@ -0,0 +1,568 @@
+```yaml
+allowObjects:
+  relationship:
+    name: 'my_relationship'
+    relatedObjectAllowed: true
+```
+
+### AuthConfig
+
+The notion of roles will be removed from AuthConfig. In particular, the `allow_role_emulation_by` key will be replaced by a `allow_session_emulation_if` key which specifies a condition which needs to evaluate to true for the request to allow emulating session variables.
+
+No `x-hasura-` session variables will now be required in the webhook response or in JWT claims. Users can choose arbitrary names for session variables.
+
+### Implementation
+
+The above proposal has major implications for engine, the `lang-graphql` crate in particular. Currently, `lang-graphql` allows constructing a single graphql schema with information about all roles embedded into the schema and allows query validation / IR generation to happen in the context of a particular role only. However, with the above proposal, there is no notion of roles anymore. So, `lang-graphql` will need to move to an architecture where when trying to access a graphql field, instead of looking up an allowed set of roles, it can evaluate an arbitrary expression that evaluates to a boolean. The details of this architecture are out of scope for this RFC.
+
+#### Milestones
+
+- Milestone 1: Refactor the internal engine implementation to support the new permissions system without any metadata changes. Internally, the role-based definitions would be mapped to the new system.
+- Milestone 2: Make the metadata changes for the new system with the following primitives:
+  - Conditional activation of rules. Conditions support a basic set of operators (`equals`, `isNull`, and logical operators at least).
+  - Value expressions can either be literals or session variables only.
+  - Support only field permissions / model select predicates / argument presets. No input validation.
+  - Composability of rules.
+- New features (in no particular order):
+  - Reusing relationship permissions
+  - Arguments input validation
+  - Calling external systems
+  - Accessing sub-fields in value expressions
+  - Advanced boolean operators
+
+## FAQs
+
+1. **Will existing metadata continue to be work?**  
+    
+    Yes, existing metadata (using the `v1` version of permissions objects) will continue to work and trivially maps to the new system where the condition is an equality check on the `x-hasura-role` session variable.
+
+2. **Does this subsume the inherited roles feature in v2?**
+    
+    Yes, this is the intended replacement for inherited roles in v2. So, instead of defining multiple base roles and an inherited you would simply define multiple rules which can activated by arbitrary conditions.
+
+    Note, however, that cell-level authorization will not be supported by this system, since you define type (column) permissions and model (row) permissions independently.
+
+3. **Are any changes required to the NDC protocol?**
```

Comment:
But if row and column permissions are disjoint, would you not want to push down the cell-level predicates? Or are you proposing that we redact data on the engine side.

That would be simpler but result in more network overhead.

I think it'd be worth spelling this out here.